File: swift/Sources/Bulwark/Spotlight.swift
// Spotlighting: random-nonce delimiting, data-marking, and base64 isolation.
// Based on spotlighting (Hines et al., Microsoft, 2024).

import Foundation

public let defaultMarker = "▁"  // ▁ LOWER ONE EIGHTH BLOCK — reads as a space marker
public let defaultTag = "untrusted_content"

private func randomHex(_ nBytes: Int = 9) -> String {
    var rng = SystemRandomNumberGenerator()  // cryptographically secure
    var s = ""
    s.reserveCapacity(nBytes * 2)
    for _ in 0..<nBytes {
        s += String(format: "%02x", UInt8.random(in: 0...255, using: &rng))
    }
    return s
}

public func makeNonce(_ nBytes: Int = 9) -> String { randomHex(nBytes) }

public func delimit(_ text: String, nonce: String? = nil, tag: String = defaultTag) -> (wrapped: String, nonce: String) {
    let n = nonce ?? makeNonce()
    let open = "<\(tag) data-nonce=\"\(n)\">"
    let close = "</\(tag) data-nonce=\"\(n)\">"
    return ("\(open)\n\(text)\n\(close)", n)
}

public func datamark(_ text: String, marker: String = defaultMarker) -> String {
    text.replacingOccurrences(of: " ", with: marker)
}

public func encodeBase64(_ text: String) -> String {
    Data(text.utf8).base64EncodedString()
}

public struct SpotlightOptions {
    public var methods: [String]
    public var nonce: String?
    public var marker: String
    public var tag: String

    public init(methods: [String] = ["delimit"], nonce: String? = nil, marker: String = defaultMarker, tag: String = defaultTag) {
        self.methods = methods
        self.nonce = nonce
        self.marker = marker
        self.tag = tag
    }
}

public func spotlight(_ text: String, options: SpotlightOptions = SpotlightOptions()) -> SpotlightResult {
    var applied: [String] = []
    var content = text
    var usedMarker: String? = nil
    var base64Encoded = false

    if options.methods.contains("base64") {
        content = encodeBase64(content)
        base64Encoded = true
        applied.append("base64")
    } else if options.methods.contains("datamark") {
        content = datamark(content, marker: options.marker)
        usedMarker = options.marker
        applied.append("datamark")
    }

    let d = delimit(content, nonce: options.nonce, tag: options.tag)
    applied.append("delimit")

    return SpotlightResult(content: d.wrapped, nonce: d.nonce, methods: applied, marker: usedMarker, base64Encoded: base64Encoded)
}

File: python/src/bulwark/spotlight.py
"""Spotlighting: make untrusted content unmistakably data.

Three composable transforms: ``delimit`` wraps content in a random-nonce
boundary a fake closing tag can't forge; ``datamark`` replaces spaces with a
marker; ``base64`` encodes the content as an opaque blob. Based on spotlighting
(Hines et al., Microsoft, 2024).
"""

from __future__ import annotations

import base64
import secrets
from typing import List, Optional, Sequence

from .types import SpotlightResult

DEFAULT_MARKER = "▁"  # ▁ LOWER ONE EIGHTH BLOCK — rare in prose, reads as a space marker
DEFAULT_TAG = "untrusted_content"


def make_nonce(n_bytes: int = 9) -> str:
    return secrets.token_hex(n_bytes)


def delimit(text: str, *, nonce: Optional[str] = None, tag: str = DEFAULT_TAG) -> "tuple[str, str]":
    """Wrap ``text`` in a nonce-tagged boundary. Returns (wrapped, nonce)."""
    nonce = nonce or make_nonce()
    open_tag = f'<{tag} data-nonce="{nonce}">'
    close_tag = f'</{tag} data-nonce="{nonce}">'
    return f"{open_tag}\n{text}\n{close_tag}", nonce


def datamark(text: str, *, marker: str = DEFAULT_MARKER) -> str:
    """Replace spaces with ``marker`` so the whole span reads as data."""
    return text.replace(" ", marker)


def encode_base64(text: str) -> str:
    return base64.b64encode(text.encode("utf-8")).decode("ascii")


def spotlight(
    text: str,
    *,
    methods: Sequence[str] = ("delimit",),
    nonce: Optional[str] = None,
    marker: str = DEFAULT_MARKER,
    tag: str = DEFAULT_TAG,
) -> SpotlightResult:
    """Apply the requested spotlight transforms in a safe order.

    ``methods`` may contain any of ``"delimit"``, ``"datamark"``, ``"base64"``.
    ``base64`` and ``datamark`` are mutually exclusive (base64 wins); ``delimit``
    is always applied even if omitted, because the boundary nonce is what the
    output validator checks for leakage.
    """
    applied: List[str] = []
    content = text
    used_marker: Optional[str] = None
    base64_encoded = False

    if "base64" in methods:
        content = encode_base64(content)
        base64_encoded = True
        applied.append("base64")
    elif "datamark" in methods:
        content = datamark(content, marker=marker)
        used_marker = marker
        applied.append("datamark")

    content, used_nonce = delimit(content, nonce=nonce, tag=tag)
    applied.append("delimit")

    return SpotlightResult(
        content=content,
        nonce=used_nonce,
        methods=applied,
        marker=used_marker,
        base64_encoded=base64_encoded,
    )
